### ISSUE TITLE
Suggestion: Implement RSA-OAEP padding in software

### ISSUE THREAD

**Issue by mouse07410** [OpenSC/OpenSC#2363]
### Problem Description

Attempts to decrypt an RSA OAEP-encrypted file fails, because `pkcs11-tool` fails to  figure what to send to the card.

The card does not do RSA-OAEP in hardware.

### Proposed Resolution

Either ensure OAEP is done in software when the card doesn't to in "on-board", or document in the `pkcs11-tool.man` page that OAEP mechanism works only with cards that do it in hardware. 

### Steps to reproduce

```
$ echo 7920c159696dae51198d103dc98e2bb51f36aedadc84aa0b9efa681c89e8606c | xxd -r -p -c 200 | openssl pkeyutl -engine pkcs11 -keyform engine -encrypt -pubin -inkey "pkcs11:model=PKCS%2315%20emulated;manufacturer=piv_II;id=%03;object=KEY%20MAN%20pubkey;type=public" -pkeyopt rsa_padding_mode:oaep -out /tmp/derive.72971.key.enc
$
$$ pkcs11-tool --decrypt --mgf MGF1-SHA384 --hash-algorithm SHA384 --salt-len -1 --id 03 -i /tmp/derive.72971.key.enc -o /tmp/derive.72971.pass.dec
Using slot 0 with a present token (0x0)
Logging in to "xxxxxxxxxxxxxxx".
Please enter User PIN: 
Using decrypt algorithm RSA-X-509
error: The hash-algorithm is applicable only to RSA-PKCS-OAEP mechanism
Aborting.
$ pkcs11-tool --decrypt -m RSA-PKCS-OAEP --mgf MGF1-SHA384 --hash-algorithm SHA384 --salt-len -1 --id 03 -i /tmp/derive.72971.key.enc -o /tmp/derive.72971.pass.dec
Using slot 0 with a present token (0x0)
Logging in to "xxxxxxxxxxxxx".
Please enter User PIN: 
Using decrypt algorithm RSA-PKCS-OAEP
OAEP parameters: hashAlg=SHA384, mgf=MGF1-SHA384, source_type=0, source_ptr=0x0, source_len=0
error: PKCS11 function C_DecryptInit failed: rv = CKR_MECHANISM_INVALID (0x70)
Aborting.

```

**Comment by dengert** [comment]
This may be a problem with opensc, not pkcs11-tool. Using opensc-0.22.0-rc1
```
 doug@XUbuntu-20:/opt/ossl-1.1.1k/bin$ ./pkcs11-tool -M
Using slot 0 with a present token (0x0)
Supported mechanisms:
  SHA-1, digest
  SHA224, digest
  SHA256, digest
  SHA384, digest
  SHA512, digest
  MD5, digest
  RIPEMD160, digest
  GOSTR3411, digest
  ECDSA, keySize={256,384}, hw, sign, verify, EC OID, EC uncompressed
  ECDSA-SHA1, keySize={256,384}, sign, verify
  ECDSA-SHA224, keySize={256,384}, sign, verify
  ECDSA-SHA256, keySize={256,384}, sign, verify
  ECDSA-SHA384, keySize={256,384}, sign, verify
  ECDSA-SHA512, keySize={256,384}, sign, verify
  ECDH1-COFACTOR-DERIVE, keySize={256,384}, hw, derive, EC OID, EC uncompressed
  ECDH1-DERIVE, keySize={256,384}, hw, derive, EC OID, EC uncompressed
  RSA-X-509, keySize={1024,3072}, hw, decrypt, sign, verify
  RSA-PKCS, keySize={1024,3072}, hw, decrypt, sign, verify
  SHA1-RSA-PKCS, keySize={1024,3072}, sign, verify
  SHA224-RSA-PKCS, keySize={1024,3072}, sign, verify
  SHA256-RSA-PKCS, keySize={1024,3072}, sign, verify
  SHA384-RSA-PKCS, keySize={1024,3072}, sign, verify
  SHA512-RSA-PKCS, keySize={1024,3072}, sign, verify
  MD5-RSA-PKCS, keySize={1024,3072}, sign, verify
  RIPEMD160-RSA-PKCS, keySize={1024,3072}, sign, verify
  RSA-PKCS-PSS, keySize={1024,3072}, hw, sign, verify
  SHA1-RSA-PKCS-PSS, keySize={1024,3072}, sign, verify
  SHA224-RSA-PKCS-PSS, keySize={1024,3072}, sign, verify
  SHA256-RSA-PKCS-PSS, keySize={1024,3072}, sign, verify
  SHA384-RSA-PKCS-PSS, keySize={1024,3072}, sign, verify
  SHA512-RSA-PKCS-PSS, keySize={1024,3072}, sign, verify
```
In other words, OpenSC does not know how to simulate RSA-PKCS-OAEP with a token that only does RSA-X-509 and pkcs11-tool does not try and simulate it, expecting the PKCS11 module will do it. 

```
doug@XUbuntu-20:/opt/ossl-1.1.1k/bin$ echo 7920c159696dae51198d103dc98e2bb51f36aedadc84aa0b9efa681c89e8606c | xxd -r -p -c 200 | ./openssl pkeyutl -engine pkcs11 -keyform engine -encrypt -pubin -inkey "pkcs11:model=PKCS%2315%20emulated;manufacturer=piv_II;id=%03;object=KEY%20MAN%20pubkey;type=public" -pkeyopt rsa_padding_mode:oaep -out /tmp/derive.72971.key.enc
engine "pkcs11" set.
doug@XUbuntu-20:/opt/ossl-1.1.1k/bin$ ls -l /tmp/derive.72971.key.enc
-rw-r--r-- 1 doug doug 256 Jul  1 16:17 /tmp/derive.72971.key.enc
doug@XUbuntu-20:/opt/ossl-1.1.1k/bin$ ./pkcs11-tool --decrypt -m RSA-PKCS-OAEP --mgf MGF1-SHA384 --hash-algorithm SHA384 --salt-len -1 --id 03 -i /tmp/derive.72971.key.enc -o /tmp/derive.72971.pass.dec
Using slot 0 with a present token (0x0)
Logging in to "Test Cardholder".
Please enter User PIN:
Using decrypt algorithm RSA-PKCS-OAEP
OAEP parameters: hashAlg=SHA384, mgf=MGF1-SHA384, source_type=0, source_ptr=(nil), source_len=0
error: PKCS11 function C_DecryptInit failed: rv = CKR_MECHANISM_INVALID (0x70)
Aborting.
```

Do you have any token or pkcs11 module that can do RSA-PKCS-OAEP? i.e. `pkcs11-tool -M` lists RSA-PKCS-OAEP?

**Comment by dengert** [comment]
See: https://github.com/OpenSC/OpenSC/blob/master/src/libopensc/padding.c#L504

git blame padding.c
3a3a465e6 (Jakub Jelen         2019-08-13 18:05:55 +0200 504)           /* TODO emulate the OAEP decryption */

**Comment by mouse07410** [comment]
> OpenSC does not know how to simulate RSA-PKCS-OAEP with a token that only does RSA-X-509 and pkcs11-tool does not try and simulate it, expecting the PKCS11 module will do it.

I see. This is probably the reason. Yubikey doesn't do OAEP on the card. OpenSSL with libp11+OpenSC does OAEP fine, because OpenSSL (or libp11? don't remember) simulates OAEP in software...

> Do you have any token or pkcs11 module that can do RSA-PKCS-OAEP? i.e. `pkcs11-tool -M` lists RSA-PKCS-OAEP?

I haven't tried it with that token, but yes - YubiHSM2 does do OAEP in hardware:
```
$ pkcs11-tool --module /usr/local/lib/pkcs11/yubihsm_pkcs11.dylib -M
Using slot 0 with a present token (0x0)
Supported mechanisms:
  RSA-PKCS, keySize={2048,4096}, hw, decrypt, sign, verify
  SHA1-RSA-PKCS, keySize={2048,4096}, hw, sign, verify
  SHA256-RSA-PKCS, keySize={2048,4096}, hw, sign, verify
  SHA384-RSA-PKCS, keySize={2048,4096}, hw, sign, verify
  SHA512-RSA-PKCS, keySize={2048,4096}, hw, sign, verify
  RSA-PKCS-PSS, keySize={2048,4096}, hw, sign, verify
  SHA1-RSA-PKCS-PSS, keySize={2048,4096}, hw, sign, verify
  SHA256-RSA-PKCS-PSS, keySize={2048,4096}, hw, sign, verify
  SHA384-RSA-PKCS-PSS, keySize={2048,4096}, hw, sign, verify
  SHA512-RSA-PKCS-PSS, keySize={2048,4096}, hw, sign, verify
  RSA-PKCS-KEY-PAIR-GEN, keySize={2048,4096}, hw, generate_key_pair
  ECDSA-KEY-PAIR-GEN, keySize={224,521}, hw, generate_key_pair, EC F_P, EC OID, EC uncompressed
  SHA-1-HMAC, keySize={1,512}, hw, sign, verify
  GENERIC-SECRET-KEY-GEN, keySize={1,1024}, hw, generate
  SHA256-HMAC, keySize={1,512}, hw, sign, verify
  SHA384-HMAC, keySize={1,1024}, hw, sign, verify
  SHA512-HMAC, keySize={1,1024}, hw, sign, verify
  ECDSA, keySize={224,521}, hw, sign, verify, EC F_P, EC OID, EC uncompressed
  ECDSA-SHA1, keySize={224,521}, hw, sign, verify, EC F_P, EC OID, EC uncompressed
  ECDH1-DERIVE, keySize={224,521}, hw, derive, EC F_P, EC OID, EC uncompressed
  RSA-PKCS-OAEP, keySize={2048,4096}, hw, decrypt
  mechtype-0xD9554204, keySize={128,256}, hw, encrypt, decrypt, wrap, unwrap
  ECDSA-SHA256, keySize={224,521}, hw, sign, verify, EC F_P, EC OID, EC uncompressed
  ECDSA-SHA384, keySize={224,521}, hw, sign, verify, EC F_P, EC OID, EC uncompressed
  ECDSA-SHA512, keySize={224,521}, hw, sign, verify, EC F_P, EC OID, EC uncompressed
  SHA-1, digest
  SHA256, digest
  SHA384, digest
  SHA512, digest
```

**Comment by frankmorgner** [comment]
If I understand correctly, you wish to have RSA-OAEP padding in software so that you can use it for your Yubikey. That's a nice idea, but OpenSC (`opensc-pkcs11`) is neither broken nor under-documented in this regard.

Please feel free to submit an additional PR that adds support for RSA-OAEP padding if the token doesn't support it in hardware.

**Comment by popovec** [comment]
Try to use PR #2475

**Comment by Jakuje** [comment]
This was solved with #2475. Closing. The "label" extension is tracked in separate PR.
